Support Spigot/Paper 1.21.3 and 1.21.4

jacob1 opened the pull request:
This PR adds support for 1.21.3 (bundles of bravery update) and 1.21.4 (The Garden Awakens). It's my first time doing a plugin update like this, but I tested it on my server and it seems to be working. For 1.21.4, it just gets it running, it doesn't support rendering the new pale garden / wood blocks.

The biggest difference in this compared to base 1.21 is I removed async chunk loading on paper (#3725). Paper did some massive overhauls of chunk loading on their end (PaperMC/Paper@730882fca9cf , PaperMC/Paper@7bd22b1835af). I think they did this prior to 1.21, but it wasn't caught, because it falls back to sync chunk loading if the classes aren't there. I only noticed because Mojang did their own big refactors to chunk loading in 1.21.2.
It would be worth keeping that ... but unless I had a lot of guidance I think that's way outside what I'm capable of redoing. It was already broken since 1.21 anyway.

I checked the mappings docs to get the right mappings. I found a few mistakes in the older helper classes, mostly inconsequential. I went back and fixed those too.

Tested on Spigot 1.21.4, Spigot 1.21.3, Paper 1.21.3, and Spigot 1.21. Let me know of any changes / additionals / removals you want.

### Jar Download
**Edit2:** This PR is merged, so I deleted my jar download. If the official build still isn't released, use the [dev builds](https://dynmap.us/builds/dynmap/), which support the new blocks now too.

jacob1 committed:
Spigot / Paper 1.21.3 support

jacob1 committed:
Spigot 1.21.3 Support: Fix block states

Leaves fix was applied to 1.21 as well. getMaterial() was removed in 1.20 so has been removed from that version onwards as well.

Brokkonaut commented:
You modified the helpers for old versions (for example replace ((NBTTagByte)val).h() with ((NBTTagByte)val).i()), was that intended?

jacob1 commented:
Yes. When I [looked it up](https://mappings.dev/1.21.3/net/minecraft/nbt/ByteTag.html), .h() was `getAsShort`, and .i() was `getAsByte`. That method is supposed to get a byte, not a short, so I fixed it and also all the other versions it affected.

That one is definitely inconsequential though, so I can remove it if it causes too much noise. It's not the only change I made to old versions though.

Brokkonaut commented:
I think it is good that you fixed it, I was just asking because that was not explicitly mentioned in the commit msg or pr.

jacob1 commented:
Yeah, I didn't mention that one, only the other two. As for the other two:

Fixed DynmapBlockState.isLeaves. In 1.21 it was using RAILS instead of LEAVES. Seems to be used for some lighting calculations.

Removed DynmapBlockState.setMaterial call in 1.20 onwards. The function it was using to get the material was removed. It's been using getSoundType() since then (obviously incorrect). As best I could tell this might be used in some connected texture pack code, but I couldn't tell if the material is really needed.

SilviodaSilva1972 commented:
What is missing to release the corrected version?

jacob1 commented:
It's done from my perspective, it just needs a maintainer to merge it.

You can compile it yourself using the guide in the [readme](https://github.com/webbukkit/dynmap?tab=readme-ov-file#how-to-build). I'm using it on my "live" server now and haven't run into any issues.
(I could post my .jar it but better to wait for an official build)

OeschMe commented:
Mind dropping the .jar file? I tried compiling it but build keeps failing.

Tsoccerguy3 commented:
> Mind dropping the .jar file? I tried compiling it but build keeps failing.

[Uploading Dynmap-3.7-beta-7_1.21.3_spigot.jar.zip…]()

Tsoccerguy3 commented:
[Dynmap-3.7-beta-7_1.21.3_spigot.jar.zip](https://github.com/user-attachments/files/17738419/Dynmap-3.7-beta-7_1.21.3_spigot.jar.zip)
Here is the compiled jar for 1.21.3 , Tested and working

SilviodaSilva1972 commented:
Thank you very much, grateful!

ThomasGITH commented:
Thanks a lot

Selunie commented:
It works for me too. Thank you ! ❤️

mblaze commented:
@LariOesch the problem with the build is that it is looking for `org.spigotmc:spigot:1.21.3-R0.1-SNAPSHOT` but it won't be able to find it. Usually this dependency is hosted in https://repo.mikeprimm.com/org/spigotmc/spigot/ but since it is maintained by the author of the plugin (assuming from the name), it is not yet there. Fortunately Gradle looks for dependencies in many places, including your local repository. This means that it should be enough to build Spigot for 1.21.3 with BuildTools (instructions [here](https://www.spigotmc.org/threads/spigot-bungeecord-1-21-2-1-21-3.667019/)) to get the missing dependency in your local Maven repository. At least this is what made it work for me. I'd strongly advise against downloading JARs from unofficial sources.

SomethingGeneric commented:
> @LariOesch the problem with the build is that it is looking for `org.spigotmc:spigot:1.21.3-R0.1-SNAPSHOT` but it won't be able to find it. Usually this dependency is hosted in https://repo.mikeprimm.com/org/spigotmc/spigot/ but since it is maintained by the author of the plugin (assuming from the name), it is not yet there. Fortunately Gradle looks for dependencies in many places, including your local repository. This means that it should be enough to build Spigot for 1.21.3 with BuildTools (instructions [here](https://www.spigotmc.org/threads/spigot-bungeecord-1-21-2-1-21-3.667019/)) to get the missing dependency in your local Maven repository. At least this is what made it work for me. I'd strongly advise against downloading JARs from unofficial sources.

thanks for the tip about BuildTools fixing the dependency error. Worked for me as well!

masmc05 commented:
You can easily make it now to always support async chunk loading

For chunk loading just use `CompletableFuture<NBTTagCompound> nbt = cw.getHandle().m().a.d(new ChunkCoordIntPair(chunk.x, chunk.z)).thenApply(op -> op.orElse(null));` instead of the provider's completable future.
For chunk saving just make getLoadedChunk as a method of MapChunkCache121_3, replace there `getAsyncSaveData.invoke(null, world.getHandle(), c)` with `SerializableChunkData.a(world.getHandle(), c)` and `save.invoke(null, world.getHandle(), c, o)` with `o.a()` (rename the `o` variable). Also remove the `if (!Bukkit.isPrimaryThread()) return null;` as it's not needed anymore

After that don't forget to change `isUnsafeAsync` to always false. Idk how to deal with sync methods, but you most probably can just throw `UnsupportedOperationException` as they'll become unused

jacob1 commented:
@masmc05 Thanks for the info. I sort of see where you mean. I don't have the time to try it out for a while though. If I ever have time to test and get it working I may submit another PR.

If you're willing you could also send in a PR, or just a branch.

Will that work on Spigot too? If so, that would be great. It's best not to rely on Paper's internal api anyway like the old code did.

MakeDecisionWorth commented:
@jacob1 
Seems like nether roof rendering is broken after this PR.
error log
https://pastebin.com/FSYF8nFH

custom-perspectives.txt
https://pastebin.com/RrZ1nqKK

worlds.txt
https://pastebin.com/BrHaZnaW

Teerack commented:
This didn't work for me. Here is my log https://i.imgur.com/dbj6FRi.png

jnvb commented:
[Dynmap-3.7-beta-7-fabric-1.21.3.zip](https://github.com/user-attachments/files/17942520/Dynmap-3.7-beta-7-fabric-1.21.3.zip)

Sharing my compiled version of fabric 1.21.3
I can PR if you guys want.

jacob1 committed:
Fix nether roof renders on Spigot 1.21.3

DynmapWorld.worldheight is reporting as 255 in 1.21.3, instead of 256 like it did before. (Overworld is now 319 -> 320). Unsure if this is a bug, since it seems right ...

Either way, do a Math.ceil to ensure there's enough sections even when the amount of blocks doesn't evenly line up with a 16 block section.

jacob1 commented:
> @jacob1 Seems like nether roof rendering is broken after this PR. error log https://pastebin.com/FSYF8nFH
> 
> custom-perspectives.txt https://pastebin.com/RrZ1nqKK
> 
> worlds.txt https://pastebin.com/BrHaZnaW

@shou692199 Seems like it's broken in 1.21.3, yeah. The nether's max world height is being reported as 255 now instead of 256. Unsure if this is a spigot bug or just an upstream Minecraft change. I think 255 is more accurate ... ? You can build the 0th block but not the 256th.

Anyway, I made a fix. Here's a [new jar](https://starcatcher.us/Jacob/Dynmap-3.7-beta-7-spigot.zip) if you want it. Keep the max height as 256 in custom-perspectives.txt (or else the top block won't render).
Edit: Updated this jar to remove the debug print.



> This didn't work for me. Here is my log https://i.imgur.com/dbj6FRi.png

Looks like it loaded fine to me. You can ignore the all.png error. I get that one too. It's a separate bug from before, unrelated to this PR.

shou692199 commented:
@jacob1 Also, there is something wrong with lighting. Too dark in updated render tiles.
![image](https://github.com/user-attachments/assets/02a9e0c5-dca4-447c-91e5-44a8692b7b8e)

Brokkonaut commented:
> @shou692199 Seems like it's broken in 1.21.3, yeah. The nether's max world height is being reported as 255 now instead of 256. Unsure if this is a spigot bug or just an upstream Minecraft change. I think 255 is more accurate ... ? You can build the 0th block but not the 256th.
> 
> Anyway, I made a fix. Here's a [new jar](https://starcatcher.us/Jacob/Dynmap-3.7-beta-7-spigot.zip) if you want it. Keep the max height as 256 in custom-perspectives.txt (or else the top block won't render).

At least in paper this will probably be changed back to 256 soon: https://github.com/PaperMC/Paper/pull/11675

But I think your solution is good to have it working in every implementation

jacob1 commented:
Thanks for the links. That seems like a confusing disaster to have the api differ in Spigot and Paper. I see though, the api clearly defines what should happen. I'll report it upstream to Spigot and see if it can get fixed there, or at least change the docs.

I didn't test in fabric. There is a 1.21.3 fabric PR now, which I bet is also affected. My fix should apply there too in theory.

jacob1 commented:
Reported to Spigot - https://hub.spigotmc.org/jira/browse/SPIGOT-7970
Edit: fixed now in Spigot

I tried checking fabric, but the PR doesn't compile because of all the chunk loading refactors Mojang made.

jacob1 committed:
Spigot 1.21.4 support

(ability to run only, not the ability to render any of the new pale garden blocks)

jacob1 commented:
I added support for Spigot 1.21.4

There were more nms mapping changes, so I created a new copy of the helper classes. (TileEntity.getBlockPos changed, and so did a bunch more NBT methods).

This doesn't add support for rendering the pale garden. Only letting dynmap run. If you try to render the new biome, everything just shows up black. No plans to try to add support right now. That would be a big time commitment to understand the texture_1 / models_1 format and then add all the new blocks.

luismayme commented:
I'm really noob to ask this, but I have never cloned a PR. How can I compile this?

jacob1 commented:
> I'm really noob to ask this, but I have never cloned a PR. How can I compile this?

You can clone my fork with `git clone https://github.com/jacob1/dynmap.git`.

From there, just follow the build steps in [the readme](https://github.com/webbukkit/dynmap?tab=readme-ov-file#how-to-build)

Or you can just download the jar in the first post, I put it there too :)

CosminPerRam commented:
Hey, are there any blockers on this PR? Looking forward to an official update.

DivineJimmi commented:
The update works, but pale oak is not rendering correctly. Textures are missing.
![Screenshot 2024-12-10 131732](https://github.com/user-attachments/assets/4ab8cd00-1c2b-4364-b18c-ab10470c05a6)
![Screenshot 2024-12-10 131725](https://github.com/user-attachments/assets/0b2eef3c-4487-493b-a865-4a20a7f98952)

jacob1 commented:
Yeah, all the new blocks are missing. I won't have more free time until the end of the month to check that. On my server I'm just dealing with it, it's annoying but at least it runs.

I wouldn't consider it a blocker to merging this PR, only a blocker to making a new release.

DavidDoesCode commented:
Hi @jacob1 I used your latest commit and added the textures. I got them from the 1.21.4 minecraft jar. There were other textures missing, but I didn't want to bloat up the commit with unneeded textures. 

https://github.com/DavidDoesCode/dynmap/tree/jacob1/v3.0

DavidDoesCode commented:
> Yes. When I [looked it up](https://mappings.dev/1.21.3/net/minecraft/nbt/ByteTag.html), .h() was `getAsShort`, and .i() was `getAsByte`. That method is supposed to get a byte, not a short, so I fixed it and also all the other versions it affected.
> 
> That one is definitely inconsequential though, so I can remove it if it causes too much noise. It's not the only change I made to old versions though.

I am not sure this is correct. You modified OLD version mappings. 
https://mappings.dev/1.18.1/net/minecraft/nbt/ByteTag.html
https://mappings.dev/1.21.3/net/minecraft/nbt/ByteTag.html

So if this is to be "compatible" with 1.21.3, it doesn't make sense to do.

DivineJimmi commented:
> Hi @jacob1 I used your latest commit and added the textures. I got them from the 1.21.4 minecraft jar. There were other textures missing, but I didn't want to bloat up the commit with unneeded textures.
> 
> https://github.com/DavidDoesCode/dynmap/tree/jacob1/v3.0

Can you upload a fixed .jar to test this?

jacob1 commented:
> Hi @jacob1 I used your latest commit and added the textures. I got them from the 1.21.4 minecraft jar. There were other textures missing, but I didn't want to bloat up the commit with unneeded textures.
> 
> https://github.com/DavidDoesCode/dynmap/tree/jacob1/v3.0

Cool, that's a good start. textures_1.txt and models_1.txt also need to be modified to make them show up properly, I think.


For your other comment, this is compatible with all versions of 1.21. I have separate mappings for 1.21.3 and 1.21.4. I fixed the older mappings which were using the wrong functions. I used that site and validated that everything I changed is pointing to the right one. So 1.18.1 still uses .h(), 1.21.3 uses .i().

DavidDoesCode commented:
@jacob1 - I'm struggling with gradle builds but hopefully I can get passed it soon, test, and share a jar. Thanks for your note on textures_1 and models_1 I will have a look.

DorianMAG commented:
@jacob1 
Thx for this work !
Tested with spigot 1.21.4, plugin start now

buddythugginz commented:
thankyou so much for this!

curtiplas commented:
@DorianMAG Do you mind providing the compiled jar for 1.21.4?  I'm struggling to get my build to compile.

mjstahlberg commented:
Compiled this today for `spigot 1.21.4` from the source branch of this PR: https://dl.viech.name/Dynmap-3.7-beta-7-spigot.jar

```
% md5sum Dynmap-3.7-beta-7-spigot.jar
63333b7868d677c862af042b9af58f7b  Dynmap-3.7-beta-7-spigot.jar
```

Joo200 commented:
> Removed DynmapBlockState.setMaterial call in 1.20 onwards. The function it was using to get the material was removed. It's been using getSoundType() since then (obviously incorrect). As best I could tell this might be used in some connected texture pack code, but I couldn't tell if the material is really needed.

This is also used in fabric. Imo this is wrong. It is used in CTMTexturePack to detect matching materials to neighbor blocks.

Maybe set the minecraft block type instead of the sound type.

Joo200 commented on bukkit-helper-121-3/src/main/java/org/dynmap/bukkit/helper/v121_3/BukkitVersionHelperSpigot121_3.java:
```diff
@@ -0,0 +1,452 @@
+        Block baseb = null;
+    	Iterator<IBlockData> iter = bsids.iterator();
+    	ArrayList<String> names = new ArrayList<String>();
+		while (iter.hasNext()) {
```
Please double check all files. You have many tabs instead of 4 spaces in your files.

Joo200 commented:
A small note: In #4177 I refactored build scripts and tested to use paperweight which can use mojang mappings and use the spigot obfuscations in the build process. That works fine so far.

mikeprimm commented:
Looks good - thanks!

ianmartsinkevich commented:
Hello everyone,

I got a bit tired of waiting for someone to add the new textures from 1.21.4, so I did it myself. Note that I am using the latest build of Dynmap ([download from dynmap.us](https://dynmap.us/builds/dynmap/Dynmap-3.7-beta-7-spigot.jar)) with all the latest changes and everything works fine.

**NB**! Pale moss carpet does not render correctly. Need to add a new class based on org.dynmap.hdmap.renderer.VineStateRenderer.class. Or need to write a bunch of weird stuff in texture_1.txt. I'm not ready to bother so much over a trifle :)

![look](https://github.com/user-attachments/assets/f819374e-cb70-45ae-91bc-96aaf8c5dadf)

![look2](https://github.com/user-attachments/assets/3764119e-33ba-4df7-8411-5a6e4f61934f)

![look3](https://github.com/user-attachments/assets/828c7de2-1cf8-40aa-abbe-9298cefb101c)

**Guide:**
1. Open your Dynmap .jar file as an archive using an archiver (7Zip, WinRAR...);
2. Add the content from the [Addition to texture_1.txt](https://github.com/user-attachments/files/18390123/Addition.to.texture_1.txt) file to the end of the texture_1.txt file;
3. Add the content from the [Addition to models_1.txt](https://github.com/user-attachments/files/18390124/Addition.to.models_1.txt) file to the end of the models_1.txt file;
4. Add the contained folders and files from the [Addition to texturepacks.zip](https://github.com/user-attachments/files/18390315/Addition.to.texturepacks.zip) archive to the root path.

To be honest, I don’t really know how to use github and don’t really want to delve into it. If someone wants, they can push a commit and add all this stuff...

Hope this helps you

---
By the way, I noticed that none of the hanging signs are rendered. Literally none, not the oak ones, not the birch ones, not the others. And there is no line for cherry_hanging_sign in textures_1.txt. But that's worth saving for another pull request.

ianmartsinkevich commented:
@jacob1
Would you like to add this to the PR? :)

jacob1 commented:
Awesome! But actually this PR was already merged, so I can't edit it.

I think you should open a new PR with your changes. I'll put your changes on my server though to also test them out. It isn't too hard to make a fork and edit the textures / models files, but I think you'll have to make a local clone to add all the pngs.

mprimmRFCode commented:
@ianmartsinkevich I scooped up the linked changes, and have added them - should be in dev builds in next hour or so

ZockiRR commented:
> You can easily make it now to always support async chunk loading
> 
> For chunk loading just use `CompletableFuture<NBTTagCompound> nbt = cw.getHandle().m().a.d(new ChunkCoordIntPair(chunk.x, chunk.z)).thenApply(op -> op.orElse(null));` instead of the provider's completable future. For chunk saving just make getLoadedChunk as a method of MapChunkCache121_3, replace there `getAsyncSaveData.invoke(null, world.getHandle(), c)` with `SerializableChunkData.a(world.getHandle(), c)` and `save.invoke(null, world.getHandle(), c, o)` with `o.a()` (rename the `o` variable). Also remove the `if (!Bukkit.isPrimaryThread()) return null;` as it's not needed anymore
> 
> After that don't forget to change `isUnsafeAsync` to always false. Idk how to deal with sync methods, but you most probably can just throw `UnsupportedOperationException` as they'll become unused

I opened a PR for that: #4188
